### PR TITLE
fix: Initialize teams earlier to avoid panic.

### LINKED ISSUE #9726: [Bug]: panic the teams engine is nil
### Problem encountered

Fairground has stopped with panic

### Observed behaviour

All nodes: validators and non-validators stopped

### Expected behaviour

It should not happen

### Steps to reproduce

```bash
1. Submit `xyz` transaction
2. With this payload
3. Query `abc` API
4. Observe the error
```

### Software version

v0.73.0-preview.11

### Failing test

_No response_

### Jenkins run

_No response_

### Configuration used

```shell
Fairground
```

### Relevant log output

```shell
Oct 11 09:04:54 n01.testnet.vega.rocks visor[2470203]: 2023-10-11T09:04:54.521Z        ERROR        tendermint        consensus/state.go:732        CONSENSUS FAILURE!!!        {"err": "runtime error: invalid memory address or nil pointer dereference", "stack": "goroutine 575 [running]:
runtime/debug.Stack()
	/opt/hostedtoolcache/go/1.20.8/x64/src/runtime/debug/stack.go:24 +0x65
github.com/tendermint/tendermint/consensus.(*State).receiveRoutine.func2()
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/consensus/state.go:732 +0x4c
panic({0x3265d00, 0x5f29160})
	/opt/hostedtoolcache/go/1.20.8/x64/src/runtime/panic.go:884 +0x213
code.vegaprotocol.io/vega/core/teams.(*SnapshottedEngine).GetAllPartiesInTeams(0x40?, 0xc000100000?)
	<autogenerated>:1 +0x1d
code.vegaprotocol.io/vega/core/execution/common.(*MarketActivityTracker).getPartiesInScope(0x3230960?, 0xc0337ef6b0?)
	/home/runner/work/vega/vega/core/execution/common/market_activity_tracker.go:521 +0xcf
code.vegaprotocol.io/vega/core/execution/common.(*MarketActivityTracker).CalculateMetricForIndividuals(0xc000e3db00, 0xc0222456c0)
	/home/runner/work/vega/vega/core/execution/common/market_activity_tracker.go:540 +0x14d
code.vegaprotocol.io/vega/core/banking.(*Engine).dispatchRequired(0xc000598ea0?, 0xc00178c740?)
	/home/runner/work/vega/vega/core/banking/recurring_transfers.go:187 +0x47
code.vegaprotocol.io/vega/core/banking.(*Engine).distributeRecurringTransfers(0xc000598ea0, {0x412bd88, 0xc0318bf950}, 0x26b7)
	/home/runner/work/vega/vega/core/banking/recurring_transfers.go:285 +0x1e39
code.vegaprotocol.io/vega/core/banking.(*Engine).OnEpoch(0xc000598ea0, {0x412bd88, 0xc0318bf950}, {0x26b7, {0x2b624354, 0xedcb84e23, 0x0}, {0x2b624354, 0xedcb85c33, 0x0}, ...})
	/home/runner/work/vega/vega/core/banking/engine.go:276 +0xcd
code.vegaprotocol.io/vega/core/epochtime.(*Svc).notify(0xc0005c97a0, {0x412bd88, 0xc0318bf950}, {0x26b7, {0x2b624354, 0xedcb84e23, 0x0}, {0x2b624354, 0xedcb85c33, 0x0}, ...})
	/home/runner/work/vega/vega/core/epochtime/service.go:201 +0xe5
code.vegaprotocol.io/vega/core/epochtime.(*Svc).OnTick(0xc0005c97a0, {0x412bd88, 0xc0318bf950}, {0x1?, 0x4146728?, 0x0?})
	/home/runner/work/vega/vega/core/epochtime/service.go:131 +0x718
code.vegaprotocol.io/vega/core/vegatime.(*Svc).notify(...)
	/home/runner/work/vega/vega/core/vegatime/service.go:104
code.vegaprotocol.io/vega/core/vegatime.(*Svc).SetTimeNow(0xc000e3d980, {0x412bd88, 0xc0318bf950}, {0x40?, 0x30fef00?, 0x0?})
	/home/runner/work/vega/vega/core/vegatime/service.go:77 +0x24f
code.vegaprotocol.io/vega/core/processor.(*App).OnBeginBlock(_, {{0xc039ff7360, 0x20, 0x20}, {{0xb, 0x1}, {0xc0348ec220, 0x1c}, 0xe6d079, {0x1e3840de, ...}, ...}, ...})
	/home/runner/work/vega/vega/core/processor/abci.go:836 +0x747
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).BeginBlock(_, {{0xc039ff7360, 0x20, 0x20}, {{0xb, 0x1}, {0xc0348ec220, 0x1c}, 0xe6d079, {0x1e3840de, ...}, ...}, ...})
	/home/runner/work/vega/vega/core/blockchain/abci/abci.go:46 +0x75
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*appW).BeginBlock(_, {{0xc039ff7360, 0x20, 0x20}, {{0xb, 0x1}, {0xc0348ec220, 0x1c}, 0xe6d079, {0x1e3840de, ...}, ...}, ...})
	/home/runner/work/vega/vega/cmd/vega/commands/node/app_wrapper.go:64 +0x55
github.com/tendermint/tendermint/abci/client.(*localClient).BeginBlockSync(_, {{0xc039ff7360, 0x20, 0x20}, {{0xb, 0x1}, {0xc0348ec220, 0x1c}, 0xe6d079, {0x1e3840de, ...}, ...}, ...})
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/abci/client/local_client.go:280 +0x118
github.com/tendermint/tendermint/proxy.(*appConnConsensus).BeginBlockSync(_, {{0xc039ff7360, 0x20, 0x20}, {{0xb, 0x1}, {0xc0348ec220, 0x1c}, 0xe6d079, {0x1e3840de, ...}, ...}, ...})
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/proxy/app_conn.go:81 +0x55
github.com/tendermint/tendermint/state.execBlockOnProxyApp({0x4129b30?, 0xc00b0a8848}, {0x413d7b0, 0xc01de37140}, 0xc0137f63c0, {0x414f200, 0xc00af39158}, 0xe6d078?)
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/state/execution.go:307 +0x51d
github.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock(_, {{{0xb, 0x1}, {0xc010cc5ba8, 0x7}}, {0xc006914e20, 0x1c}, 0x1, 0xe6d078, {{0xc02f0fedc0, ...}, ...}, ...}, ...)
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/state/execution.go:140 +0x171
github.com/tendermint/tendermint/consensus.(*State).finalizeCommit(0xc00124dc00, 0xe6d079)
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/consensus/state.go:1661 +0xafd
github.com/tendermint/tendermint/consensus.(*State).tryFinalizeCommit(0xc00124dc00, 0xe6d079)
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/consensus/state.go:1570 +0x2ff
github.com/tendermint/tendermint/consensus.(*State).enterCommit.func1()
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/consensus/state.go:1505 +0xaa
github.com/tendermint/tendermint/consensus.(*State).enterCommit(0xc00124dc00, 0xe6d079, 0x0)
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/consensus/state.go:1543 +0xccf
github.com/tendermint/tendermint/consensus.(*State).addVote(0xc00124dc00, 0xc014275040, {0xc019adb560, 0x28})
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/consensus/state.go:2164 +0x18dc
github.com/tendermint/tendermint/consensus.(*State).tryAddVote(0xc00124dc00, 0xc014275040, {0xc019adb560?, 0xc00178fc08?})
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/consensus/state.go:1962 +0x2c
github.com/tendermint/tendermint/consensus.(*State).handleMsg(0xc00124dc00, {{0x41077e0, 0xc019398698}, {0xc019adb560, 0x28}})
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/consensus/state.go:861 +0x3ff
github.com/tendermint/tendermint/consensus.(*State).receiveRoutine(0xc00124dc00, 0x0)
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/consensus/state.go:768 +0x3f0
created by github.com/tendermint/tendermint/consensus.(*State).OnStart
	/home/runner/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.29-patch.1/consensus/state.go:379 +0x12d
"}
```

### PR DESCRIPTION
**Pull request by karlem**
Close #9726

### CODE REVIEW

File: core/protocol/all_services.go [hunk 1]
```diff
@@ -257,6 +257,14 @@ func newServices(
 	svcs.epochService.NotifyOnEpoch(svcs.topology.OnEpochEvent, svcs.topology.OnEpochRestore)
 	svcs.epochService.NotifyOnEpoch(stats.OnEpochEvent, stats.OnEpochRestore)
 
+	// The team engine is used to know the team a party belongs to. The computation
+	// of the referral program rewards requires this information. Since the team
+	// switches happen when the end of epoch is reached, it needs to be one of the
+	// last services to register on epoch update, so the computation is made based
+	// on the team the parties belonged to during the epoch and not the new one.
+	svcs.teamsEngine = teams.NewSnapshottedEngine(svcs.epochService, svcs.broker, svcs.timeService)
+	svcs.snapshotEngine.AddProviders(svcs.teamsEngine)
```
Comment (ValentinTrinque) [inline core/protocol/all_services.go:266]: This changes has to be done in the integration test setup as well.

**Review by ValentinTrinque — CHANGES_REQUESTED** [review]

File: core/protocol/all_services.go [hunk 2]
```diff
@@ -257,6 +257,14 @@ func newServices(
 	svcs.epochService.NotifyOnEpoch(svcs.topology.OnEpochEvent, svcs.topology.OnEpochRestore)
 	svcs.epochService.NotifyOnEpoch(stats.OnEpochEvent, stats.OnEpochRestore)
 
+	// The team engine is used to know the team a party belongs to. The computation
+	// of the referral program rewards requires this information. Since the team
+	// switches happen when the end of epoch is reached, it needs to be one of the
+	// last services to register on epoch update, so the computation is made based
+	// on the team the parties belonged to during the epoch and not the new one.
```
Comment (ValentinTrinque) [inline core/protocol/all_services.go:264]: The content of this comment is important, and that change must respect it.

**Review by ValentinTrinque — CHANGES_REQUESTED** [review]

File: core/protocol/all_services.go [hunk 3]
```diff
@@ -257,6 +257,14 @@ func newServices(
...
+	svcs.teamsEngine = teams.NewSnapshottedEngine(svcs.epochService, svcs.broker, svcs.timeService)
```
Comment (ze97286) [inline core/protocol/all_services.go:265]: the way you changes it isn't gonna work because of the comment right above it...

Comment (ze97286) [reply core/protocol/all_services.go:265]: you could move the wiring of the epoch to here then you control when it is done and ensure it happens last or towards the end.

File: core/protocol/all_services.go (hunk 1 above)
Comment (jeremyletang) [reply core/protocol/all_services.go:266]: I've done it before, without realising it was borked in the core too.

File: core/protocol/all_services.go (hunk 2 above)
Comment (ze97286) [reply core/protocol/all_services.go:264]: so glad we have this comment :)

File: core/protocol/all_services.go (hunk 1 above)
Comment (ValentinTrinque) [reply core/protocol/all_services.go:266]: It's just so we have a consistent order of calling the callback between the integraton test and the core.

**Review by ValentinTrinque — APPROVED** [review]

**Review by ValentinTrinque — DISMISSED** [review]

**Review by asiaznik — DISMISSED** [review]

**Review by ValentinTrinque — DISMISSED** [review]

File: core/protocol/all_services.go [hunk 4]
```diff
@@ -257,6 +257,9 @@ func newServices(
 	svcs.epochService.NotifyOnEpoch(svcs.topology.OnEpochEvent, svcs.topology.OnEpochRestore)
 	svcs.epochService.NotifyOnEpoch(stats.OnEpochEvent, stats.OnEpochRestore)
 
+	svcs.teamsEngine = teams.NewSnapshottedEngine(svcs.epochService, svcs.broker, svcs.timeService)
```
Comment (ze97286) [inline core/protocol/all_services.go:260]: then you don't need the epoch service in the teams engine anymore no?

**Review by ze97286 — DISMISSED** [review]

**Review by ze97286 — DISMISSED** [review]

**Review by jeremyletang — APPROVED** [review]

**Review by jeremyletang — DISMISSED** [review]

**Review by jeremyletang — DISMISSED** [review]

**Pull request merged**